# UK Driving Accidents Database

### <font color=green>*Udacity Data Engineering Nanodegree - Capstone Project*</font>

#### Project Summary
This project aims to build a database that analysts at the Department for Transport, of the United Kingdom Government will be able to use.
The database will cover information pertaining to road traffic accidents recorded, with accompanying information, such as location data.
The database will be designed in a star-schema format, where accident records will provide the FACT table, and accompanying reference information will form the DIMENSION tables.

The project follows the follow steps:
* Step 1: Scope the Project and Gather Data
* Step 2: Explore and Assess the Data
* Step 3: Define the Data Model
* Step 4: Run ETL to Model the Data
* Step 5: Complete Project Write Up

-----------------

#### Project Environment
This project is managed through a conda environment. Where required, you may need to `conda` or `pip` install specific packages into your own evironment. <br>
`pip install` commands that can be run from the notebook to the terminal will be place below, but commented out. You can un-comment these and install as required to your environment.


In [1]:
## install commands if needed
#!pip install ipython-sql
#!pip install pandas
#!pip install json
#!pip install os
#!pip install glob
#!pip install zipfile
#!pip install boto3
#!pip install botocore
#!pip install configparser 
#!pip install psycopg2
#!pip install requests

In [2]:
# Imports 
#import kaggle
import pandas as pd
import json 
import os
import glob 
import zipfile 
import boto3
from botocore.client import ClientError 
import configparser
import psycopg2 
import requests
from AWS_S3_Functions import checkS3bucket, buildS3bucket, loadToS3


# so we can explore all columns on any pandas DF printed out 
pd.set_option('display.max_columns', None) 
pd.set_option('display.max_colwidth', None) 

----------------

## Step 1: Project Scope & Data Gather
#### Scope 
The requirements of this project are to create an ETL pipeline which builds ready to use analytical tables for analysts within the Department for Transport in the UK Government. This is so analysts can produce MI & insights into national, regional & constituency road traffic accidents. It will also allow analysts to investigate the links between Driving Test centres, average income, local police forces, and other variables.

##### Project Steps
- Source Data from online sources. Download these data to local storage as part of the project folder.
- Import each data source, explore, clean and export back to a new location for holding clean data files.
- Create an S3 bucket on AWS, to store these newly cleaned files for use within an ETL pipeline. Large, cheap and easily accessible storage is available via AWS S3.
- Next, a redshift cluster will be created, which will contain staging tables (loading the data from S3) before extracting, transforming and loading that data into a series of analytical tables (star schema format) which will be used by the analysts in the DoT of the UK Government. 
- Finally the staging tables are then dropped.
- Some DQ queries are ran against the new database analytical tables, to check for DQ constraints and that tables have loaded as expected.
- This project finally finishes by cleaning up the different resources created; removing the redshift cluster, the IAM role and the S3 objects & bucket.


Key steps of the process are configured via the **"AWS.cfg"** file, in order to allow the pipeline to be easily modified for individual use, with personal details, removing the need to modify multiple lines of code.


#### The Data Sources
This project makes use of datasets that are sourced from Kaggle (though originating from gov.uk open source data) and the free to access website `www.doogal.co.uk` which provides UK Location data. Below will cover some more information for each data source to be used in this project. <br>
The project provides code that will collect the data from the relevant web-pages and store locally in the project working directory.

#### <font color=green>** UK Road Safety: Traffic Accidents & Vehicles Data **</font>

The following data is sourced via Kaggle. <br>
The data originally come from the Open Data website of the UK Government, as detailed in the Kaggle description. <br> 
That site is linked here: <https://data.gov.uk/dataset/cb7ae6f0-4be6-4935-9277-47e5ce24a11f/road-safety-data> 

As described on the kaggle post:

These datasets features a wide range of information regarding traffic accident in the UK, such as vehicle details, accident location, road details, weather details & much more
- *Accident Information* : every line within the file represents a unique traffic accident (identified by the AccidentIndex column), featuring various properties related to the accident as columns. Date range: 2005-2017
- *Vehicle Information* : every line in the file represents the involvement of a unique vehicle in a unique traffic accident, featuring various vehicle and passenger properties as columns. Date range: 2004-2016 

In [3]:
# Use the kaggle API to download both "Accident" & "Vehicle" CSV files - run via a command line call
!kaggle datasets download -d tsiaras/uk-road-safety-accidents-and-vehicles 


  0%|          | 0.00/142M [00:00<?, ?B/s]
  1%|▏         | 2.00M/142M [00:00<00:11, 13.0MB/s]
  3%|▎         | 4.00M/142M [00:00<00:11, 13.1MB/s]
  4%|▍         | 6.00M/142M [00:00<00:11, 12.6MB/s]
  6%|▌         | 8.00M/142M [00:00<00:10, 13.1MB/s]
  7%|▋         | 10.0M/142M [00:00<00:10, 13.4MB/s]
  8%|▊         | 12.0M/142M [00:00<00:10, 13.2MB/s]
 10%|▉         | 14.0M/142M [00:01<00:10, 13.4MB/s]
 11%|█         | 16.0M/142M [00:01<00:09, 13.7MB/s]
 13%|█▎        | 18.0M/142M [00:01<00:09, 13.7MB/s]
 14%|█▍        | 20.0M/142M [00:01<00:09, 13.9MB/s]
 15%|█▌        | 22.0M/142M [00:01<00:09, 13.8MB/s]
 17%|█▋        | 24.0M/142M [00:01<00:08, 13.9MB/s]
 18%|█▊        | 26.0M/142M [00:02<00:08, 13.8MB/s]
 20%|█▉        | 28.0M/142M [00:02<00:08, 13.9MB/s]
 21%|██        | 30.0M/142M [00:02<00:08, 13.7MB/s]
 22%|██▏       | 32.0M/142M [00:02<00:08, 13.6MB/s]
 24%|██▍       | 34.0M/142M [00:02<00:08, 13.6MB/s]
 25%|██▌       | 36.0M/142M [00:02<00:08, 13.7MB/s]
 27%|██▋       | 38.

In [4]:
# from the zip file that will have been downloaded, extract the contents
accidents_location = r"uk-road-safety-accidents-and-vehicles.zip" 

with zipfile.ZipFile(accidents_location, 'r') as zip_ref:
    zip_ref.extractall() 

#### <font color=green>** UK Driving Test Centres Data **</font>

The following data is sourced via Kaggle. <br>
The data originally come from the Open Data website of the UK Government, as detailed in the Kaggle description. <br> 
That site is linked here: <https://data.gov.uk/dataset/fe19beff-5716-4ca9-be58-027e56856b48/driving-test-centres> 

The original on the site, is in CSV format, however, the kaggle version has been created in JSON format.

This dataset contains locations of Driving Test Centres in the UK. The test centre name, postcode and latitude & longitude co-ordinates are available in this data.

In [5]:
# download data to zipped file 
!kaggle datasets download -d dancollins3dmc/uk-driving-test-centres


  0%|          | 0.00/38.2k [00:00<?, ?B/s]
100%|██████████| 38.2k/38.2k [00:00<00:00, 3.91MB/s]


In [6]:
# from the zip file that will have been downloaded, extract the contents
test_centre_location = r"uk-driving-test-centres.zip"  

with zipfile.ZipFile(test_centre_location, 'r') as zip_ref:
    zip_ref.extractall() 

#### <font color=green>** UK LSOA location data **</font>

The following data is sourced via the website Doogal. <br>
That site is linked here: <https://www.doogal.co.uk/postcodedownloads.php>

These datasets contains key location data for UK Postcodes & LSOAs (Lower Layer Super Output Areas). <br> 
The files also contains data regarding political constituency, ditrict, ward, nearest train station, police force & average income. <br> 

Each Country is accessible via CSV format through a URL.

- England : <https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=England>
- Wales : <https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=Wales>
- Scotland : <https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=Scotland>
- Northern Ireland : <https://www.doogal.co.uk/UKPostcodesCSV.ashx?Search=BT> 

<br>
This project will actually use a summarised version of these 4 postcode files, where the data will be grouped to the distinct LSOA level, and the Police Force, Avg Income & other high level location information retained. 

In [7]:
# Download all 4 postcode files to working Directory
urls = [r"https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=England",
        r"https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=Wales",   
        r"https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=Scotland",      
        r"https://www.doogal.co.uk/UKPostcodesCSV.ashx?Search=BT"
]
countries = ["England", "Wales", "Scotland", "Northern Ireland"] 
filelists = pd.DataFrame.from_dict({'URL': urls, 'Country': countries})
filelists 

URL  \
0   https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=England   
1     https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=Wales   
2  https://www.doogal.co.uk/UKPostcodesCSV.ashx?country=Scotland   
3         https://www.doogal.co.uk/UKPostcodesCSV.ashx?Search=BT   

            Country  
0           England  
1             Wales  
2          Scotland  
3  Northern Ireland

In [8]:
# loop through each row, read chunks from the URL, download into a textfile locally 
for index, row in filelists.iterrows():
    url = row['URL'] 
    response = requests.get(url, stream = True) 
    outfile = row['Country'] + ".txt" 
    textfile = open(outfile, "wb")
    print("Start Writing " + outfile)
    for chunk in response.iter_content(chunk_size=1024):
        textfile.write(chunk)

    print("Finished Writing " + outfile)
    textfile.close()  

Start Writing England.txt
Finished Writing England.txt
Start Writing Wales.txt
Finished Writing Wales.txt
Start Writing Scotland.txt
Finished Writing Scotland.txt
Start Writing Northern Ireland.txt
Finished Writing Northern Ireland.txt


In [10]:
## Read file in (via chunks to reduce out of memory errors) 
# columns to retain 
lsoa_cols = ["LSOA Code", "Country", "Constituency", "Average Income", "Police force"] 
# list of files downloaded 
txt_list = ["England.txt", "Scotland.txt", "Wales.txt", "Northern Ireland.txt"] 
# empty holder list 
all_files = [] 

for txt in txt_list:
    for chunk in pd.read_csv(txt, usecols=lsoa_cols, chunksize=10000):
        all_files.append(chunk) 

lsoa_data = pd.concat(all_files, axis=0) # appends all smaller dataframes into one larger one (vertically) 

In [11]:
# write the data back to one CSV file 
lsoa_data.to_csv("LSOA_Data.csv", sep='~', index=False) 

In [12]:
# clean up no longer required separate files
for txt in txt_list:
    os.remove(txt)
# end     

In [13]:
# some clean-up 
del lsoa_data, all_files 

---------------
## Step 2: Explore & Assess Data
#### Data Exploration - UK Driving Test Centres

Start by looking at the `UK Driving Test Centres` JSON file. Read in and explore the data.

In [14]:
test_centres = pd.read_json(r"DriveTest_Centres.json") 
test_centres.head(3) 

Name   Address line 1                 Address line 2  \
0         Aberdeen North  Cloverhill Road                Bridge of Don     
1  Aberdeen South (Cove)        Moss Road  Gateway Business Park, Nigg     
2              Aberfeldy        Town Hall                 Crieff Road      

        City             County  Postcode        Lat      Long  \
0   Aberdeen      Aberdeenshire  AB23 8FE  57.185055 -2.095251   
1   Aberdeen      Aberdeenshire  AB12 3GQ  57.088588 -2.107476   
2  Aberfeldy  Perth and Kinross  PH15 2BJ  56.618039 -3.868265   

                                           Toilets  \
0  Male, female and disabled toilets are available   
1  Male, female and disabled toilets are available   
2  Male, female and disabled toilets are available   

                                    Disabled access  \
0  You can get into the test centre in a wheelchair   
1  You can get into the test centre in a wheelchair   
2  You can get into the test centre in a wheelchair   

                                        Parking  \
0  No car park is available at the test centre    
1              Car parking spaces are available   
2                                          None   

                                                                                                                                                                                                 Other notes  \
0  For Car Tests: Please Keep to the one way system around the building and meet your examiner at the VIC waiting room on left of the DVSA building. Please also note that no practising is allowed on site.   
1                                                                                                                                                                                                       None   
2                                                                                                                                                                                                       None   

   Car Motorcycle Module 1 Motorcycle module 2 LBCCT   B+E  Taxi ADI Part 2  \
0  Car                None                None   Voc   B+E   NaN       ADI2   
1  Car                 Mc1                 Mc2  None  None   NaN       None   
2  Car                None                None  None  None   NaN       None   

  ADI Part 3  TARSID  Closing date  
0       ADI3   337.0           NaN  
1       ADI3  1878.0           NaN  
2       None   247.0           NaN

### Only some variables are of interest for the database being built
Variables of interest are: <br>
- `Name, City, County, Postcode, Lat, Long, Closing date`

In [15]:
# count the number of rows in the dataset
print("Test Centre file has " + str(len(test_centres)) + " rows") 

Test Centre file has 362 rows


In [16]:
# check for nulls in these key columns 
tc_cols = ["Name", "City", "County", "Postcode", "Lat", "Long", "Closing date"]
display(test_centres[tc_cols].isnull().sum()) 

Name              2
City              6
County            2
Postcode          2
Lat               2
Long              2
Closing date    362
dtype: int64

We can see from the results that, two of the values in the `Name` column are null. These will be dropped from the table. <br>
There also appears to be two nulls within Latitude & Longitude, which are key variables for furture joins. This would need to be dropped also. Let's check if they are all the same cases. <br>
We can also see all `Closing date` values are NULL, so no test centres are, or due to be, closed

In [17]:
tc = test_centres[tc_cols] 
tc.loc[tc['Name'].isnull()] 

Name  City County Postcode  Lat  Long  Closing date
360  None  None   None     None  NaN   NaN           NaN
361  None  None   None     None  NaN   NaN           NaN

It looks like all these cases are indeed linked. As such, these will need to be dropped.
- Cleaning Note 1: Drop cases where `Name` is NULL

In [18]:
# check for & return any duplicate rows 
tc[tc.duplicated(['Name'], keep=False)] 

Name  City County Postcode  Lat  Long  Closing date
360  None  None   None     None  NaN   NaN           NaN
361  None  None   None     None  NaN   NaN           NaN

Good. No duplicated `Names` other than the NULLs of which we were aware. <br>
Let's take a look at cases where `City` was NULL.

In [19]:
tc.loc[tc['City'].isnull()] 

Name  City           County  Postcode        Lat      Long  \
27                Barry  None  South Glamorgan  CF62 5QN  51.399694 -3.279244   
145  Goodmayes (London)  None   Greater London   IG3 9UZ  51.563820  0.110019   
208          Letchworth  None    Hertfordshire   SG6 1RF  51.978204 -0.214656   
253             Orkney   None           Orkney  KW15 1FL  58.981674 -2.972001   
360                None  None             None      None        NaN       NaN   
361                None  None             None      None        NaN       NaN   

     Closing date  
27            NaN  
145           NaN  
208           NaN  
253           NaN  
360           NaN  
361           NaN

So, except for the cases of NULL, we have 4 cases where `City` is empty. 
For these cases, we will clean them by filling the empty column ith the `County`. <br>

#### Data Clean - UK Driving Test Centres
- Remove cases where `Name` is NULL
- Fill `City` blanks with `County` column values

Write cleaned dataset back out to a new file.

In [20]:
tc_clean = tc.loc[tc['Name'].notnull()] 
print("Number of records after NULL drops: " +str(len(tc_clean))) 
# we should now see an output of 360 rows, 2 having been dropped 

# next, clean `City` column to copy over the `County` value 
tc_clean = tc_clean.copy() 
tc_clean.loc[tc_clean['City'].isnull(), 'City'] = tc_clean['County'] 
# re-run check for nulls in this column, should now be zero 
check_city = tc_clean.loc[tc_clean['City'].isnull()] 
print("Number of NULL `City` values: " + str(len(check_city))) 

Number of records after NULL drops: 360
Number of NULL `City` values: 0


In [21]:
# check if the "clean_data" directory exists, if not, create it 
if not os.path.exists("clean_data"):
    os.makedirs("clean_data")

# write clean data back to clean file 
tc_clean.to_json("clean_data/UK_Driving_Test_Centre.json", orient='records', lines=True) 

del tc_clean, tc 

#### Data Exploration - UK LSOA Data

Read in and explore the data. 

In [22]:
lsoa = pd.read_csv("LSOA_Data.csv", sep='~', usecols=lsoa_cols) 
lsoa.head(3) 

Country Constituency  LSOA Code   Police force  Average Income
0  England    St Albans  E01023743  Hertfordshire         68900.0
1  England    St Albans  E01023667  Hertfordshire         66500.0
2  England    St Albans  E01023667  Hertfordshire         66500.0

In [23]:
print("LSOA has " + str(len(lsoa)) + " rows") 
print("** ------------------------------ **")
lsoa_cols = lsoa.columns.values.tolist() 
for col in lsoa_cols:
    unique_count = lsoa[col].nunique() 
    print(col + " has " + str(unique_count) + " unique values") 

LSOA has 2652196 rows
** ------------------------------ **
Country has 4 unique values
Constituency has 650 unique values
LSOA Code has 42619 unique values
Police force has 45 unique values
Average Income has 532 unique values


Some basic checks confirm good DQ here. 4 countries, as does make up the United Kingdom, so we would expect to see in the data built here from our source. Likewise, the UK has 650 seats in the house of commons (UK parliament) which matches to the 650 `Constituency` count we can see here.

Also, as we can see `LSOA Code` has the highest frequency of distinct values. This was expected as LSOA is a quite granular location variable, used by the UK ONS for reporting small area statistics. <br> 

In [24]:
# check info() 
lsoa.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2652196 entries, 0 to 2652195
Data columns (total 5 columns):
 #   Column          Dtype  
---  ------          -----  
 0   Country         object 
 1   Constituency    object 
 2   LSOA Code       object 
 3   Police force    object 
 4   Average Income  float64
dtypes: float64(1), object(4)
memory usage: 101.2+ MB


In [25]:
# search for instances of LSOA Code being null - these will need to be removed as part of the cleaning
for col in lsoa_cols:
    null_count = lsoa.loc[lsoa[col].isnull()]
    print("Null count in " + col + " is: "+ str(len(null_count))) 

Null count in Country is: 0
Null count in Constituency is: 10332
Null count in LSOA Code is: 10332
Null count in Police force is: 10332
Null count in Average Income is: 295946


#### Data Clean - UK LSOA location data

By removing the null LSOA rows, we will also clean the District, Constituency & Police Force columns. <br> 
For missing Average Income, we will use the latest UK ONS **Median** Household Income value, as a proxy for average income. 

Cleaning Actions:
- remove the 10,332 rows with NULL `LSOA Code` 
- where `Average Income` is missing, replace with UK 2020 Median Income value according to the ONS (£29,900)  <https://www.ons.gov.uk/peoplepopulationandcommunity/personalandhouseholdfinances/incomeandwealth/bulletins/householddisposableincomeandinequality/financialyear2020> 

In [26]:
# replace NULLs in Avg Income with 29,900 
lsoa['Average Income'] = lsoa['Average Income'].fillna(29900) 

# drop rows where LSOA Code is NULL 
lsoa.dropna(subset=['LSOA Code'], axis=0, inplace=True)  # axis=0 specifies rows containing 0 dropped as per Pandas API docs

# check NULLs again 
for col in lsoa.columns.values.tolist():
    null_count = lsoa.loc[lsoa[col].isnull()]
    print("Null count in " + col + " is: "+ str(len(null_count)))

Null count in Country is: 0
Null count in Constituency is: 0
Null count in LSOA Code is: 0
Null count in Police force is: 0
Null count in Average Income is: 0


We no longer have any NULL values. <br>
The final stage is to summarise with a `groupby()` to remoave the duplication at LSOA Code level, taking the mean of the `Average Income` column 

In [27]:
groupCols = ["Country", "Constituency", "LSOA Code", "Police force"] 
lsoa_clean = lsoa.groupby(groupCols).mean().reset_index() 
lsoa_clean.sample(3) 

Country         Constituency  LSOA Code        Police force  \
33143  England          Wyre Forest  E01032469         West Mercia   
17772  England   Manchester, Gorton  E01005280  Greater Manchester   
3353   England  Boston and Skegness  E01026013        Lincolnshire   

       Average Income  
33143         31200.0  
17772         30900.0  
3353          41300.0

In [28]:
# check the number of duplicate rows in dataset - should be 0
print("The number of duplicate rows in lsoa_clean is: " + str(len(lsoa_clean[lsoa_clean.duplicated()]))) 

The number of duplicate rows in lsoa_clean is: 0


In [29]:
# check that no duplicates exist at the LSOA & Constituency level (which will act as Primary Keys) - Answer should be `False`
lsoa_clean[['LSOA Code','Constituency']].duplicated().any() 

False

In [30]:
# write clean data back to CSV file in clean_data folder 
# check if the "clean_data" directory exists, if not, create it 
if not os.path.exists("clean_data"):
    os.makedirs("clean_data")

# write clean data back to clean file 
lsoa_clean.to_csv("clean_data/UK_LSOA_Data.csv", sep='~', index=False) 

del lsoa_clean, lsoa 

#### Data Exploration - UK Accidents Information

Read in and explore the data. 

In [31]:
# create a list of the key columns of interest 
accident_cols = ["Accident_Index", "1st_Road_Class", "1st_Road_Number", "Accident_Severity", "Date", "Day_of_Week", 
                  "Did_Police_Officer_Attend_Scene_of_Accident", "Latitude", "Longitude", "LSOA_of_Accident_Location",
                  "Number_of_Casualties", "Number_of_Vehicles", "Road_Surface_Conditions", "Road_Type", "Speed_limit",
                  "Time", "Year", "Weather_Conditions" 
] 
acc_files = [] # empty list to store chunk read-ins
for chunk in pd.read_csv("Accident_Information.csv", usecols=accident_cols, chunksize=10000):
    acc_files.append(chunk) 
    

accidents = pd.concat(acc_files, axis=0) 
accidents.sample(3) 

Accident_Index 1st_Road_Class  1st_Road_Number Accident_Severity  \
1407594  2013160B02691              A            180.0           Serious   
1635882  2014950005820              A              1.0            Slight   
375650   2006920600152   Unclassified              0.0           Serious   

               Date Day_of_Week  Did_Police_Officer_Attend_Scene_of_Accident  \
1407594  2013-09-02      Monday                                          1.0   
1635882  2014-11-11     Tuesday                                          1.0   
375650   2006-01-11   Wednesday                                          1.0   

          Latitude  Longitude LSOA_of_Accident_Location  Number_of_Casualties  \
1407594  53.590495  -0.391215                 E01013270                     3   
1635882  55.928875  -3.022169                       NaN                     1   
375650   57.161759  -2.148266                       NaN                     1   

         Number_of_Vehicles Road_Surface_Conditions           Road_Type  \
1407594                   1                     Dry    Dual carriageway   
1635882                   2             Wet or damp    Dual carriageway   
375650                    1                     Dry  Single carriageway   

         Speed_limit   Time  Weather_Conditions  Year  
1407594         70.0  15:55  Fine no high winds  2013  
1635882         70.0  17:30  Fine no high winds  2014  
375650          30.0  17:45   Fine + high winds  2006

In [32]:
accidents.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2047256 entries, 0 to 2047255
Data columns (total 18 columns):
 #   Column                                       Dtype  
---  ------                                       -----  
 0   Accident_Index                               object 
 1   1st_Road_Class                               object 
 2   1st_Road_Number                              float64
 3   Accident_Severity                            object 
 4   Date                                         object 
 5   Day_of_Week                                  object 
 6   Did_Police_Officer_Attend_Scene_of_Accident  float64
 7   Latitude                                     float64
 8   Longitude                                    float64
 9   LSOA_of_Accident_Location                    object 
 10  Number_of_Casualties                         int64  
 11  Number_of_Vehicles                           int64  
 12  Road_Surface_Conditions                      object 
 13  Road_Type   

In [33]:
# count rows of data
print("Accidents has " + str(len(accidents)) + " rows") 

print("** ------------------------------------------ **")

# check for NULLs in each column 
for col in accident_cols:
    null_count = accidents.loc[accidents[col].isnull()]
    print("Null count in " + col + " is: "+ str(len(null_count)))

Accidents has 2047256 rows
** ------------------------------------------ **
Null count in Accident_Index is: 0
Null count in 1st_Road_Class is: 0
Null count in 1st_Road_Number is: 2
Null count in Accident_Severity is: 0
Null count in Date is: 0
Null count in Day_of_Week is: 0
Null count in Did_Police_Officer_Attend_Scene_of_Accident is: 278
Null count in Latitude is: 174
Null count in Longitude is: 175
Null count in LSOA_of_Accident_Location is: 144953
Null count in Number_of_Casualties is: 0
Null count in Number_of_Vehicles is: 0
Null count in Road_Surface_Conditions is: 0
Null count in Road_Type is: 0
Null count in Speed_limit is: 37
Null count in Time is: 156
Null count in Year is: 0
Null count in Weather_Conditions is: 0


Initial notes for cleaning:
- drop all rows where the `LSOA_of_Accident_Location` is NULL
- drop all rows where the `Latitude` or `Longitude` are NULL <br> 
        *This is because these fields are key in order to join additional reference data that will be provided in the database for analysis*  <br>
<br> 
- where `Time` is NULL, set a default value of midnight (00:00) 
- where `Did_Police_Officer_Attend_Scene_of_Accident` is NULL, default to "NO" (value of 2) & clean the column to remove 1/0 and replace with descriptive Y or N 
- where `Speed_limit` is null, replace with default value? Maybe national speed limit of 60MPH
- retain `Date` field, can drop `Year` & `Day_of_Week` as these will be derived from `Date` during table definitions in DB
- create a `Timestamp` variable that uses both `Date` & `Time`



In [34]:
accidents["1st_Road_Class"].value_counts() 

A               926729
Unclassified    603938
B               258076
C               174953
Motorway         78071
A(M)              5489
Name: 1st_Road_Class, dtype: int64

In [35]:
accidents["Did_Police_Officer_Attend_Scene_of_Accident"].value_counts()  

1.0    1638195
2.0     403424
3.0       5359
Name: Did_Police_Officer_Attend_Scene_of_Accident, dtype: int64

Additional Cleaning Notes:
- Clean the `A(M)` category into just `A` for Road Class
- rename variables to just `Road_Class` and `Road_Number` 
- Re-classify the Police Attendance into a new variable 


#### Data Clean - UK Accidents Data

In [36]:
# drop rows where LSOA, Lat or Long are NULL
accidents.dropna(subset=["Latitude", "Longitude", "LSOA_of_Accident_Location"], axis=0, inplace=True) 

# fill in NULL times with 00:00
accidents['Time'] = accidents['Time'].fillna("00:00") 

# fill in NULL speed limits with 60MPH (UK National Speed Limit) 
accidents['Speed_limit'] = accidents['Speed_limit'].fillna(60) 

# clean missing police officer attend values to default NO (aka 2)
accidents["Did_Police_Officer_Attend_Scene_of_Accident"] = accidents["Did_Police_Officer_Attend_Scene_of_Accident"].fillna(2) 
# NEW column, providing cleaner descriptive values for police attendance 
def police_attend(row):
    if row["Did_Police_Officer_Attend_Scene_of_Accident"] == 1 :
        return 'Yes' 
    elif row["Did_Police_Officer_Attend_Scene_of_Accident"] == 2 :
        return 'No'
    elif row["Did_Police_Officer_Attend_Scene_of_Accident"] == 3 :
        return 'No - Accident self reported via form' 


accidents["Police_Attended"] = accidents.apply(lambda row: police_attend(row), axis=1) 

# drop columns Year & Day_of_Week & OLD Did Police Officer attend columns 
accidents.drop(columns=["Year", "Day_of_Week", "Did_Police_Officer_Attend_Scene_of_Accident"], axis=0, inplace=True) 

# replace A(M) with just A in Road Class 
accidents["1st_Road_Class"].replace({"A(M)": "A"}, inplace=True) 

# rename columns 
accidents.rename(columns={"1st_Road_Class": "Road_Class", "1st_Road_Number": "Road_Number"}, inplace=True)  

In [37]:
accidents.info() 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1902289 entries, 0 to 2040142
Data columns (total 16 columns):
 #   Column                     Dtype  
---  ------                     -----  
 0   Accident_Index             object 
 1   Road_Class                 object 
 2   Road_Number                float64
 3   Accident_Severity          object 
 4   Date                       object 
 5   Latitude                   float64
 6   Longitude                  float64
 7   LSOA_of_Accident_Location  object 
 8   Number_of_Casualties       int64  
 9   Number_of_Vehicles         int64  
 10  Road_Surface_Conditions    object 
 11  Road_Type                  object 
 12  Speed_limit                float64
 13  Time                       object 
 14  Weather_Conditions         object 
 15  Police_Attended            object 
dtypes: float64(4), int64(2), object(10)
memory usage: 246.7+ MB


In [38]:
new_accident_cols = accidents.columns.values.tolist() 
# check for NULLs in each column 
for col in new_accident_cols:
    null_count = accidents.loc[accidents[col].isnull()]
    print("Null count in " + col + " is: "+ str(len(null_count)))

Null count in Accident_Index is: 0
Null count in Road_Class is: 0
Null count in Road_Number is: 0
Null count in Accident_Severity is: 0
Null count in Date is: 0
Null count in Latitude is: 0
Null count in Longitude is: 0
Null count in LSOA_of_Accident_Location is: 0
Null count in Number_of_Casualties is: 0
Null count in Number_of_Vehicles is: 0
Null count in Road_Surface_Conditions is: 0
Null count in Road_Type is: 0
Null count in Speed_limit is: 0
Null count in Time is: 0
Null count in Weather_Conditions is: 0
Null count in Police_Attended is: 0


We no longer have any NULLs, and now have all the variable names required. <br>
Write the cleaned dataset to a new CSV file in the clean_data folder 

In [39]:
# write clean data back to CSV file in clean_data folder 
# check if the "clean_data" directory exists, if not, create it 
if not os.path.exists("clean_data"):
    os.makedirs("clean_data")

# write clean data back to clean file 
accidents.to_csv("clean_data/UK_Accidents_Information.csv", sep='~', index=False) 

del accidents 

#### Data Exploration - UK Vehicles Information

Read in and explore the data. 

In [40]:
# create a list of the key columns of interest 
vehicle_cols = [
    "Accident_Index", "Age_Band_of_Driver", "Age_of_Vehicle", "make", "model", "Engine_Capacity_.CC.",
    "Sex_of_Driver", "Vehicle_Type", "Vehicle_Manoeuvre", "X1st_Point_of_Impact"
]
veh_files = [] # empty list to store chunk read-ins
# change encoding as utf-8 throws an error 
for chunk in pd.read_csv("Vehicle_Information.csv", sep=',', encoding="ISO-8859-1", usecols=vehicle_cols, chunksize=10000):
    veh_files.append(chunk) 
    

vehicles = pd.concat(veh_files, axis=0) 
vehicles.sample(3) 

Accident_Index Age_Band_of_Driver  Age_of_Vehicle  \
832558   2010521005771            26 - 35             9.0   
1622690  2014460238038            46 - 55            13.0   
1434246  2013440357636            36 - 45             2.0   

         Engine_Capacity_.CC.     make             model Sex_of_Driver  \
832558                 1984.0    VOLVO      S60 T S AUTO     Not known   
1622690                1242.0     FORD  FIESTA FREESTYLE        Female   
1434246                1248.0  HYUNDAI       I20 COMFORT        Female   

         Vehicle_Manoeuvre Vehicle_Type X1st_Point_of_Impact  
832558       Turning right          Car                Front  
1622690  Going ahead other          Car                Front  
1434246         Moving off          Car                Front

In [41]:
vehicles.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2177205 entries, 0 to 2177204
Data columns (total 10 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Accident_Index        object 
 1   Age_Band_of_Driver    object 
 2   Age_of_Vehicle        float64
 3   Engine_Capacity_.CC.  float64
 4   make                  object 
 5   model                 object 
 6   Sex_of_Driver         object 
 7   Vehicle_Manoeuvre     object 
 8   Vehicle_Type          object 
 9   X1st_Point_of_Impact  object 
dtypes: float64(2), object(8)
memory usage: 166.1+ MB


In [42]:
# check for NULLs in each column 
for col in vehicle_cols:
    null_count = vehicles.loc[vehicles[col].isnull()]
    print("Null count in " + col + " is: "+ str(len(null_count)))

Null count in Accident_Index is: 0
Null count in Age_Band_of_Driver is: 0
Null count in Age_of_Vehicle is: 358149
Null count in make is: 110845
Null count in model is: 325331
Null count in Engine_Capacity_.CC. is: 265861
Null count in Sex_of_Driver is: 0
Null count in Vehicle_Type is: 0
Null count in Vehicle_Manoeuvre is: 0
Null count in X1st_Point_of_Impact is: 0


In [43]:
# check for any duplicates within the Accident_Index 
dupe_check = vehicles['Accident_Index'].duplicated().any() 
dupe_check 

True

So we can see that Accident Index is indeed duplicated within the vehicles data. This is likely as we multiple vehicles involved in accidents as times. <br> 
Let's check assumption by finding a duplicate `Accident_Index` record

In [44]:
duplicateIds = vehicles[vehicles.duplicated(['Accident_Index'], keep=False)] 
duplicateIds.head(5) 

Accident_Index Age_Band_of_Driver  Age_of_Vehicle  Engine_Capacity_.CC.  \
2  200401BS00003            26 - 35             4.0                 998.0   
3  200401BS00003            66 - 75             NaN                   NaN   
4  200401BS00004            26 - 35             1.0                 124.0   
5  200401BS00004            36 - 45            10.0                1781.0   
9  200401BS00013            26 - 35             4.0                2300.0   

               make                  model Sex_of_Driver  Vehicle_Manoeuvre  \
2            NISSAN  MICRA CELEBRATION 16V          Male      Turning right   
3  LONDON TAXIS INT         TXII GOLD AUTO          Male  Going ahead other   
4           PIAGGIO              VESPA ET4          Male  Going ahead other   
5        VOLKSWAGEN                    NaN          Male             U-turn   
9        VOLKSWAGEN                GOLF V5        Female      Turning right   

                 Vehicle_Type X1st_Point_of_Impact  
2                         109                Front  
3                         109                Front  
4  Motorcycle 125cc and under                Front  
5                         109              Offside  
9                         109              Offside

So, the duplicates aren't an issue, as more than one vehicle can be involved in an accident. Now, let's look at the vehicle type, because we are going to create a vehicle dimension table as part of the data model, storing details and linking via an ID.

In [45]:
# check the counts on vehicle type
vehicles["Vehicle_Type"].value_counts() 

Car                                      1528628
Van / Goods 3.5 tonnes mgw or under       117427
109                                        82920
Bus or coach (17 or more pass seats)       76757
Motorcycle over 500cc                      71472
Motorcycle 125cc and under                 61600
Goods 7.5 tonnes mgw and over              55426
Taxi/Private hire car                      43781
Pedal cycle                                38904
Motorcycle 50cc and under                  22415
Motorcycle over 125cc and up to 500cc      20960
Goods over 3.5t. and under 7.5t            18236
Other vehicle                              13994
106                                         7568
Agricultural vehicle                        6018
Minibus (8 - 16 passenger seats)            5900
Goods vehicle - unknown weight              1876
108                                         1334
Motorcycle - unknown cc                      741
Mobility scooter                             502
Data missing or out 

So, we can see quite a wide-range of entries in the vehicle type column. There are certainly some cases here where:
- A broad category of similar descriptions
- not actually a road vehicle

With these categories in mind, we will perform some group cleaning on this field. <br> 

- re-classify 109, 108 & 106 to `Car`
- group all the motorcycle types to one `Motorcycle` group 
- rename `Van / Goods ...` to just `Van` 
- group all other goods vehicles into one `Goods Vehicle` group 
- rename `Bus or Coach` to `Bus/Coach` 
- rename `Minibus (8-16 passenger seats)` to just `Minibus` 
- change `Data Missing` to `N/A` 

Place into a new "clean" column. We can then drop the old one & rename 

#### Data Clean - UK Vehicles Data

In [46]:
# start by setting the current column values as default, then override as needed per descriptions above 
vehicles["Vehicle_Type_Clean"] = vehicles["Vehicle_Type"] 

# now run through creating new values per list above 
vehicles.loc[vehicles.Vehicle_Type.isin(["106", "108", "109"]), 'Vehicle_Type_Clean'] = 'Car' 

vehicles.loc[vehicles.Vehicle_Type \
    .isin(["Motorcycle - unknown cc", "Motorcycle 50cc and under", "Motorcycle 125cc and under", 
            "Motorcycle over 125cc and up to 500cc", "Motorcycle over 500cc"]), 'Vehicle_Type_Clean'] = 'Motorcycle'

vehicles.loc[vehicles.Vehicle_Type.isin(["Van / Goods 3.5 tonnes mgw or under"]), 'Vehicle_Type_Clean'] = 'Van' 

vehicles.loc[vehicles.Vehicle_Type.isin(["Bus or coach (17 or more pass seats)"]), 'Vehicle_Type_Clean'] = 'Bus/Coach' 

vehicles.loc[vehicles.Vehicle_Type.isin(["Minibus (8 - 16 passenger seats)"]), 'Vehicle_Type_Clean'] = 'Minibus' 

vehicles.loc[vehicles.Vehicle_Type.isin(["Goods vehicle - unknown weight", "Goods over 3.5t. and under 7.5t", "Goods 7.5 tonnes mgw and over"]), 'Vehicle_Type_Clean'] = 'Goods Vehicle' 

vehicles.loc[vehicles.Vehicle_Type.isin(["Data missing or out of range"]), 'Vehicle_Type_Clean'] = 'N/A' 

In [47]:
# check the counts on vehicle type again after changes 
vehicles["Vehicle_Type_Clean"].value_counts() 

Car                      1620450
Motorcycle                177188
Van                       117427
Bus/Coach                  76757
Goods Vehicle              75538
Taxi/Private hire car      43781
Pedal cycle                38904
Other vehicle              13994
Agricultural vehicle        6018
Minibus                     5900
Mobility scooter             502
N/A                          401
Ridden horse                 224
Electric motorcycle           78
Tram                          43
Name: Vehicle_Type_Clean, dtype: int64

In [48]:
# drop old vehicle type column, then rename new one 
vehicles.drop('Vehicle_Type', axis=1, inplace=True) 
vehicles.rename({'Vehicle_Type_Clean': 'Vehicle_Type'}, axis=1, inplace=True) 
vehicles.rename({'Engine_Capacity_.CC.': 'Engine_Capacity_CC'}, axis=1, inplace=True) 
vehicles.head(3) 

Accident_Index Age_Band_of_Driver  Age_of_Vehicle  Engine_Capacity_CC  \
0  200401BS00001            26 - 35             3.0              1588.0   
1  200401BS00002            26 - 35             NaN                 NaN   
2  200401BS00003            26 - 35             4.0               998.0   

     make                  model Sex_of_Driver  Vehicle_Manoeuvre  \
0   ROVER         45 CLASSIC 16V          Male  Going ahead other   
1     BMW                     C1          Male  Going ahead other   
2  NISSAN  MICRA CELEBRATION 16V          Male      Turning right   

  X1st_Point_of_Impact Vehicle_Type  
0                Front          Car  
1                Front          Car  
2                Front          Car

In [49]:
# write clean data back to CSV file in clean_data folder 
# check if the "clean_data" directory exists, if not, create it 
if not os.path.exists("clean_data"):
    os.makedirs("clean_data")

# write clean data back to clean file 
vehicles.to_csv("clean_data/UK_Vehicles_Information.csv", sep='~', index=False) 

del vehicles 

Remove all the original files, retaining just the clean copies to be used as part of the Data Pipeline 

In [50]:
rem = ["Accident_Information.csv", "Vehicle_Information.csv", "DriveTest_Centres.json", "LSOA_Data.csv", "uk-driving-test-centres.zip", "uk-road-safety-accidents-and-vehicles.zip"]
for item in rem:
    os.remove(item) 

-----------
### Step 3: Define the Data Model
#### 3.1 Conceptual Data Model
Conceptual Data Model for the UK Road Accidents Database for the United Kingdom Government, Department of Transport

<img src="graphics\Data_Model2.PNG" width="900" height="540">

#### 3.2 Mapping Out Data Pipelines
#### ETL Process Flow
<img src="graphics\Data_Pipeline.PNG" width="1200" height="350"> 

----------
### Step 4: Run Pipelines to Model the Data 
#### 4.1 Create the data model
Build the data pipelines to create the data model. <br>



#### Stage 1:
Start by using some imported functions from `AWS_S3_Functions` which will set up an S3 bucket on AWS, then load our files to storage there.

In [51]:
# use the config parser to collect details from the AWS.cfg file which will feed into the creation of AWS S3 services 
config = configparser.ConfigParser() 
config.read_file(open('AWS.cfg')) 

KEY = config.get('AWS','KEY')
SECRET = config.get('AWS','SECRET')
useRegion = config.get('BUCKET','REGION')
useBucketName = config.get('BUCKET','NAME') 

# create S3 AWS client 
aws_s3_client = boto3.client('s3',
                            region_name = useRegion,
                            aws_access_key_id = KEY,
                            aws_secret_access_key = SECRET) 

In [52]:
# check if the bucket already exists 
bucket_exists = checkS3bucket(s3_client=aws_s3_client, bucketName=useBucketName, clientError=ClientError) 
# If bucket does not currently exist, attempt to create it 
if bucket_exists == 'No':
    buildIt = buildS3bucket(s3_client=aws_s3_client, bucketName=useBucketName, clientError=ClientError, aws_region=useRegion) 
    print(buildIt) 

else: 
    print("Bucket already exists - no creation required") 

This bucket does not currently exist
S3 bucket successfully created


In [53]:
# With the bucket no successfully created, load data to it (note, this can take up to 10 minutes ... so grab a coffee!)  
for file in glob.glob("clean_data\*"):
    fileName = file.split('\\')[1] 
    loaded_file = loadToS3(s3_client=aws_s3_client, bucketName=useBucketName, clientError=ClientError, bucketKey='project_data/', loadFile=file, fileName=fileName) 
    print(loaded_file) 

Successfully loaded UK_Accidents_Information.csv to dmc-021593-capstone
Successfully loaded UK_Driving_Test_Centre.json to dmc-021593-capstone
Successfully loaded UK_LSOA_Data.csv to dmc-021593-capstone
Successfully loaded UK_Vehicles_Information.csv to dmc-021593-capstone


#### Stage 2:
Now that the data is stored in AWS S3, we will set up a redshift cluster & db, which will contain the staging tables and end user analyst tables. <br>
We will need to create some AWS clients to work with. This will involve creating a specified IAM role to create a Redshift cluster & have read-only access to the S3 bucket created

In [54]:
aws_redshift_client = boto3.client('redshift',
                            region_name = useRegion,
                            aws_access_key_id = KEY,
                            aws_secret_access_key = SECRET) 

aws_iam_client = boto3.client('iam',
                            region_name = useRegion,
                            aws_access_key_id = KEY,
                            aws_secret_access_key = SECRET)

aws_ec2_resource = boto3.resource('ec2',
                            region_name=useRegion,
                            aws_access_key_id = KEY,
                            aws_secret_access_key = SECRET)

In [55]:
# Create the IAM role choosing a name & policy 
roleName = config.get('AWS','IAM_ROLE')
policy = "arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess" 
try:
    print("Creating a new IAM Role ...") 
    dwhRole = aws_iam_client.create_role(
        Path='/',
        RoleName=roleName,
        Description = "Allows Redshift clusters to call AWS services on your behalf",
        AssumeRolePolicyDocument=json.dumps(
            {'Statement': [{'Action': 'sts:AssumeRole',
               'Effect': 'Allow',
               'Principal': {'Service': 'redshift.amazonaws.com'}}],
             'Version': '2012-10-17'})
    )    
    print("IAM role " + str(roleName) + " successfully created")
except Exception as e:
    print("IAM role creation failed") 
    print(e) 

# Attach the policy to the IAM role 
try:
    print("Attaching Policy to IAM role ...")
    aws_iam_client.attach_role_policy(RoleName=roleName,
                        PolicyArn=policy)['ResponseMetadata']['HTTPStatusCode'] 
    print("Policy attached to IAM role") 
except Exception as e:
    print("IAM role policy attach failed") 
    print(e) 

# Collect the ARN details of the IAM role 
roleArn = aws_iam_client.get_role(RoleName=roleName)['Role']['Arn']

Creating a new IAM Role ...
IAM role etldev successfully created
Attaching Policy to IAM role ...
Policy attached to IAM role


With the IAM role now created, launch a AWS redshift cluster. <br> 
The details for the clusters configuration can be collected from the AWS.cfg file <br>
#### Create Redshift Cluster 

In [56]:
try:
    print("Creating Redhsift cluster " + str(config.get('DWH','DWH_CLUSTER_IDENTIFIER')) + " ...") 
    response = aws_redshift_client.create_cluster(        
        #HW
        ClusterType=config.get('DWH','DWH_CLUSTER_TYPE'),
        NodeType=config.get('DWH','DWH_NODE_TYPE'),
        NumberOfNodes=int(config.get('DWH','DWH_NUM_NODES')),

        #Identifiers & Credentials
        DBName=config.get('DWH','DWH_DB'),
        ClusterIdentifier=config.get('DWH','DWH_CLUSTER_IDENTIFIER'),
        MasterUsername=config.get('DWH','DWH_DB_USER'),
        MasterUserPassword=config.get('DWH','DWH_DB_PASSWORD'),
        
        #Roles (for s3 access)
        IamRoles=[roleArn]  
    )
    print("Cluster " + str(config.get('DWH','DWH_CLUSTER_IDENTIFIER')) + " has been created") 
except Exception as e:
    print("Cluster creation has failed") 
    print(e) 

Creating Redhsift cluster dmc0293UKaccidents ...
Cluster dmc0293UKaccidents has been created


In [58]:
# we can use the following function to check for an active cluster - ClusterStatus must change from `creating` to `available` in order to be ready for use 
def prettyRedshiftProps(props):
    pd.set_option('display.max_colwidth', None)
    keysToShow = ["ClusterIdentifier", "NodeType", "ClusterStatus", "MasterUsername", "DBName", "Endpoint", "NumberOfNodes", 'VpcId']
    x = [(k, v) for k,v in props.items() if k in keysToShow]
    return pd.DataFrame(data=x, columns=["Key", "Value"])

myClusterProps = aws_redshift_client.describe_clusters(ClusterIdentifier=config.get('DWH','DWH_CLUSTER_IDENTIFIER'))['Clusters'][0]
prettyRedshiftProps(myClusterProps)  

Key  \
0  ClusterIdentifier   
1           NodeType   
2      ClusterStatus   
3     MasterUsername   
4             DBName   
5           Endpoint   
6              VpcId   
7      NumberOfNodes   

                                                                                           Value  
0                                                                             dmc0293ukaccidents  
1                                                                                      dc2.large  
2                                                                                      available  
3                                                                                        dbadmin  
4                                                                                   uk_accidents  
5  {'Address': 'dmc0293ukaccidents.cms6ffzqbc3y.us-west-2.redshift.amazonaws.com', 'Port': 5439}  
6                                                                                   vpc-22e0c85a  
7                                                                                              4

In [ ]:
# Collects the Endpoint & ARN 
DWH_ENDPOINT = myClusterProps['Endpoint']['Address']
DWH_ROLE_ARN = myClusterProps['IamRoles'][0]['IamRoleArn']
print("DWH_ENDPOINT :: ", DWH_ENDPOINT)
print("DWH_ROLE_ARN :: ", DWH_ROLE_ARN)  

In [60]:
# Opens an incoming TCP port to access the cluster 
# try:
#     useVpc = myClusterProps['VpcId'] 
#     vpc = aws_ec2_resource.Vpc(id=useVpc)
#     defaultSg = list(vpc.security_groups.all())[0]
#     print(defaultSg)
#     try:
#         defaultSg.authorize_ingress(
#             GroupName=defaultSg.group_name,
#             CidrIp='0.0.0.0/0',
#             IpProtocol='TCP',
#             FromPort=int(config.get('DWH','DWH_PORT')),
#             ToPort=int(config.get('DWH','DWH_PORT'))
#         )
#     except:
#         pass 
#     print("TCP port open") 
# except Exception as e:
#     print("Opening TCP Port failed") 
#     print(e) 

write the Endpoint (aka Host) and the ARN out to the AWS.cfg file

In [61]:
cfg_file = 'AWS.cfg' 
config = configparser.ConfigParser()
config.read(cfg_file)

def set_config(sec, attr, value):
    config.set(sec, attr, value)
    with open(cfg_file, 'w') as configfile:
        config.write(configfile)


set_config('CLUSTER', 'HOST', DWH_ENDPOINT)
set_config('CLUSTER', 'ROLE_ARN', DWH_ROLE_ARN) 

#### Stage 3:
So, with data now stored in an S3 bucket, and a redshift cluster & database created, we can run the main stages of the ETL process. <br>
This will be the create tables stage, which will build a series of empty tables, before then actually loading those tables with data. <br>
This pattern will be done in two parts:
- First, staging tables. Dropped if pre-existing, then created, then copied into 
- Second analytics tables, Dropped if pre-existing, then created, then inserted into from the staging tables 

Finally, we will "clean up" the database by removing the unrequired staging tables once the analytics tables are built.

In [62]:
%run staging_tables.py 

Now the staging tables have been built, we can create and load data for the analytical tables via the next script 

In [64]:
%run analytics_tables.py 

Now run a clean up script, where all staging tables are removed after the analytics tables have been created 

In [65]:
%run drop_staging_tables.py 

#### 4.2 Data Quality Checks

1 - Run a check to ensure that no duplicates exist for the combination of the `Accident_Index` & `Vehicle_Number` column from the Accidents FACT table <br>
2 - Run a check to ensure that the Vehicles DIMENSION table contains *Car Van, Motorcycle & Bus/Coach* within the `Vehicle_Type` column, as they are they main categories that should exist <br>
3 - Run a check to ensure no NULL values in `Latitude` & `Longitude` within the Test Centres DIMENSION table <br>
4 - Run a check to ensure that there are no duplicates in the `LSOA code` & `Constituency` combinations within the Locations DIMENSION table <br> 

In [66]:
# load the magic sql module for the notebook 
%load_ext sql 
#%reload_ext sql

# set connection details to redshift db 
usr = config.get('DWH','dwh_db_user')
pw = config.get('DWH','dwh_db_password') 
ep = config.get('CLUSTER','host') 
port = config.get('DWH','dwh_port') 
db = config.get('DWH','dwh_db') 
conn_string="postgresql://{}:{}@{}:{}/{}".format(usr, pw, ep, port, db)
#print(conn_string)
%sql $conn_string 

Data Quality check 1 - `duplicate_records` should = 0 

In [67]:
%%sql
SELECT COUNT(a.*) as duplicate_records
FROM (SELECT Accident_Index, Vehicle_Number, count(*) as count_ FROM accidents GROUP BY Accident_Index, Vehicle_Number) as a
WHERE a.count_ > 1 ;

 * postgresql://dbadmin:***@dmc0293ukaccidents.cms6ffzqbc3y.us-west-2.redshift.amazonaws.com:5439/uk_accidents
1 rows affected.


[(0,)]

Data Quality check 2 - Vehicles DIMENSION table DOES contain *Car Van, Motorcycle & Bus/Coach* within the `Vehicle_Type` column 

In [68]:
%%sql 
SELECT DISTINCT vehicle_type FROM Vehicles 
WHERE vehicle_type IN ('Car', 'Van', 'Motorcycle', 'Bus/Coach') ;

 * postgresql://dbadmin:***@dmc0293ukaccidents.cms6ffzqbc3y.us-west-2.redshift.amazonaws.com:5439/uk_accidents
4 rows affected.


[('Bus/Coach',), ('Motorcycle',), ('Car',), ('Van',)]

Data Quality check 3 - ensure no NULL values in `Latitude` & `Longitude` within the Test Centres DIMENSION table <br> 
count result should equal 0 

In [69]:
%%sql 
SELECT COUNT(*) as NULL_Records
FROM driving_test_centres 
WHERE latitude IS NULL or longitude IS NULL ;

 * postgresql://dbadmin:***@dmc0293ukaccidents.cms6ffzqbc3y.us-west-2.redshift.amazonaws.com:5439/uk_accidents
1 rows affected.


[(0,)]

Data Quality check 4 - check to ensure that there are no duplicates in the `LSOA` code within the Locations DIMENSION table 

In [70]:
%%sql
SELECT COUNT(a.*) as duplicate_records
FROM (SELECT lsoa, constituency, count(*) as count_ FROM locations GROUP BY lsoa, constituency) as a
WHERE a.count_ > 1 ;

 * postgresql://dbadmin:***@dmc0293ukaccidents.cms6ffzqbc3y.us-west-2.redshift.amazonaws.com:5439/uk_accidents
1 rows affected.


[(0,)]

In [71]:
# close the connection
connections = %sql -l
[c.session.close() for c in connections.values()] 

[None]

#### 4.3 Data dictionary 

#### Accidents
| *Column* | *Type* | *Description* |
| :- | :- | :- |
| Accident_Index | VARCHAR | An alphanumeric key to indicate a Traffic Accident Record. Combined with the `Vehicle_Number`, it creates a **<font color='red'>Primary Key</font>** |
| Vehicle_Number | INT | A numeric key to signal which vehicle involved in the accident this represents. Combined with the `Accident_Index`, it creates a **<font color='red'>Primary Key</font>** |
| Vehicle_Key | INT | A unique key to link to the `Vehicles` Table. Details on the vehicle involved in the accident |
| DateTime_Key | VARCHAR | A unique key to link to the `DateTime` Table. Date & time of the accident |
| Latitude | VARCHAR | The latitude co-ordinates of where the accident occurred |
| Longitude | VARCHAR | The longitude co-ordinates of where the accident occurred |
| LSOA | VARCHAR | The Lower Layer Super Output Area (a geographic hierarchy in the UK) of where the accident occurred |
| First_Road_Class | VARCHAR | The classification of the road the accident occurred on. This, combined with the `First_Road_Number` provides the road identification (except for unlclassified roads) |
| First_Road_Number | VARCHAR | The number of the road the accident occurred on. This, combined with `First_Road_Class` provides the road identification (except for unlclassified roads) |
| Road_Surface_Conditions | VARCHAR | Road surface conditions at accident site |
| Weather_Conditions | VARCHAR | Weather conditions at accident site |
| Police_Attended | VARCHAR | Did police officer attend scene at accident site |
| Number_of_Casualties | VARCHAR | Number of casualties from accident |
| Number_of_Vehicles | VARCHAR | Number of vehicles involved in accident |
| Accident_Severity | VARCHAR | The severity of the accident |
| Age_Band_of_Driver | VARCHAR | Age band of driver of this vehicle, in this accident |
| Sex_of_Driver | VARCHAR | Sex of driver of this vehicle, in this accident |
| Age_of_Vehicle | VARCHAR | Age of vehicle involved in accident |
| First_Point_of_Impact | VARCHAR | First point of impact of vehicle involved in accident |

#### Vehicles
| *Column* | *Type* | *Description* |
| :- | :- | :- |
| Vehicle_Key | INT | A unique key to link the Vehicles DIM table to the `Accidents` table **<font color='red'>Primary Key</font>** |
| Vehicle_Type | VARCHAR | The type of vehicle involved in the accident. This can be values such as a Car, Van, Motorcycle, Bus, or in some cases, non-vehicles where involved in accidents, i.e a Horse |
| Make | VARCHAR | The vehicle manufacturer, for example Nissan or Audi |
| Model | VARCHAR | The vehicle model, for example Juke (Nissan) or Q5 (Audi) |
| Engine_Capacity | VARCHAR | The CC of the engine of the vehicle. Not Applicable for non-vehicles |

#### Driving_Test_Centres 
| *Column* | *Type* | *Description* |
| :- | :- | :- |
| Latitude | VARCHAR | The latitude co-ordinates of where the test centre is based **<font color='red'>Primary Key</font>** |
| Longitude | VARCHAR | The longitude co-ordinates of where the test centre is based **<font color='red'>Primary Key</font>** |
| Name | VARCHAR | Name of the Test Centre **<font color='red'>Primary Key</font>** |
| City | VARCHAR | City the Test Centre is in |
| Postcode | VARCHAR | Postcode of the Test Centre Address |

#### Locations 
| *Column* | *Type* | *Description* |
| :- | :- | :- |
| LSOA | VARCHAR | Lower Layer Super Output Area (a geographic hierarchy in the UK) providing a mapping key for granular locations for reporting purposes via UK ONS **<font color='red'>Primary Key</font>** |
| Constituency | VARCHAR | The parliamentary name for a specific area/boundary which makes up one of 650 within the UK Government **<font color='red'>Primary Key</font>** |
| Country | VARCHAR | England, Scotland, Wales or Northern Ireland - the countries making up the United Kingdom |
| Police_Force | VARCHAR | The name of the relevant police force operating in the LSOA area |
| Avg_Income | DECIMAL(20.2) | The average income of an individual within the LSOA area |

#### DateTime 
| *Column* | *Type* | *Description* |
| :- | :- | :- |
| DateTime_Key | VARCHAR | A unique key to link the DateTime DIM table to the `Accidents` table **<font color='red'>Primary Key</font>**| 
| Date | DATE | The date of the accident |
| Time | VARCHAR | The time of the accident |
| Year | VARCHAR | The year of the accident |
| Month | VARCHAR | The month of the accident |
| Day | VARCHAR | The day of the accident |
| Weekday | VARCHAR | The weekday of the accident |

------------------------
### <font color='Red'>**Clean Up ETL resources**</font> 

- Delete Cluster
- Delete IAM role 
- Delete S3 bucket
- Delete Clean Data stored locally

In [72]:
# delete cluster 
try:
    aws_redshift_client.delete_cluster(ClusterIdentifier=config.get('DWH','DWH_CLUSTER_IDENTIFIER'),  SkipFinalClusterSnapshot=True) 
    print("Cluster Deleted") 
except Exception as e:
    print("Cluster Deletion has failed")
    print(e) 

Cluster Deleted


In [73]:
# remove IAM role created 
try:
    aws_iam_client.detach_role_policy(RoleName=config.get('AWS','IAM_ROLE'), PolicyArn="arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess")
    print("IAM Policy removed") 
    print("------------------")
    aws_iam_client.delete_role(RoleName=config.get('AWS','IAM_ROLE')) 
    print("IAM Role deleted") 
    print("------------------")

except Exception as e:
    print("Failed to delete IAM role") 
    print(e) 

IAM Policy removed
------------------
IAM Role deleted
------------------


In [75]:
# empty S3 bucket 
try:
    BUCKET = config.get('BUCKET','NAME')
    PREFIX = 'project_data/'

    response = aws_s3_client.list_objects_v2(Bucket=BUCKET, Prefix=PREFIX)

    for object in response['Contents']:
        print('Deleting', object['Key'])
        aws_s3_client.delete_object(Bucket=BUCKET, Key=object['Key']) 

except Exception as e:
    print("S3 Bucket failed to empty")
    print(e)

print("--------------------------------------------------------------") 

Deleting project_data/UK_Accidents_Information.csv
Deleting project_data/UK_Driving_Test_Centre.json
Deleting project_data/UK_LSOA_Data.csv
Deleting project_data/UK_Vehicles_Information.csv
--------------------------------------------------------------


In [80]:
aws_s3_resource = boto3.resource('s3',
                            region_name = useRegion,
                            aws_access_key_id = KEY,
                            aws_secret_access_key = SECRET) 

try:
    aws_s3_resource.Bucket(BUCKET).delete() 
    print("S3 bucket has been deleted") 

except Exception as e:
    print("S3 bucket failed to delete") 
    print(e) 

S3 bucket has been deleted


In [82]:
# remove clean data from local storage 
clean_list = [r"clean_data\UK_Accidents_Information.csv", r"clean_data\UK_Vehicles_Information.csv", r"clean_data\UK_LSOA_Data.csv", r"clean_data\UK_Driving_Test_Centre.json"]
for clean_file in clean_list:
    os.remove(clean_file) 
# end   

------------
#### Step 5: Complete Project Write Up

### <font color='green'>Tools & Technologies</font>

During the initial stages of data sourcing, exploration and cleaning, the Pandas library was chosen as the main tool. This package has a wide range of abilities. It is able to source & read in data from a variety of file formats, such as TXT, CSV & JSON used in this project, but can also access web URLs too for files stored on the web rather than a local network/folder. <br>
Pandas also proves easy to use with exploring data in its dataframe format. Column values can be explored for values, duplicates and counts. New columns or cleaned columns can be easily created and added to an existing dataframe, or summarised results can be written to a new dataframe. <br>
These cleaned/summarised dataframes can easily be written back out to different file formats, such as CSV or JSON. <br>

With regards to the ETL process, I chose to utilsie AWS S3 as the storage for the `source data`. AWS S3 offers large, cheap and easy to access storage for all files types, on the cloud. This proves beneficial as it allows easy access to other engineers with the correct access keys, provides ample storage space for a variety of file types, and can be easily connected to work with other AWS products or tools. <br> 
In order to automate as much as possible, Python to AWS SDKs were used, to create an S3 bucket and load the gathered data sources to that bucket for future use. <br>
With the bucket created, and cleaned source data stored, the next decision was to choose a way to ETL the data and place into analytical tables for future use. <br> 

The next stage of the ETL process uses the Python AWS SDK again, this time, so create a specific IAM role with S3 Read access, in order to pull the stored data. A redshift cluster is also created. <br> 
Within this cluster, we create a series of staging tables where we copy data from S3 to redshift (bulk loaded for quick copy). These staging tables prove a useful intermediate step, allowing the pipeline to access different data from each of the staging tables to build the analytical tables that analysts can use for MI reporting and querying. <br> 

Redshift has the capability to be scaled in different sizes depending on need and use case, so as large data builds when collecting a history over time, or more computational power is needed, this can be achieved. <br> 

The data model chosen, a star schema, seemed appropriate as we had key event data, the traffic accidents themselves, alongside multiple reference points. By providing dimension tables covering vehicle details (make, model, engine size), location details (areas, constituencies for Parliament Reporting), Income and Local Driving Test Centres to accident spots, a variety of different insights can be achieved. Some examples of those may be:

- Which constituencies see the most accidents? Is there also a link to low/high income areas?
- Are there driving test centres near-by? does this suggest those taking a test could be involved? are some test centres maybe passing drivers who are not quite ready? There could be some initial exploratory work to look into this
- what vehicles are most frequently found in accidents? which motorways or A roads are most affected?


### <font color='green'>Updating Data</font> 

The sourced data for this project has different intervals. <br>
The Postcode data containing LSOA codes, average incomes and police force details updates each quarter, in a release by the UK ONS. <br>
Accidents information updates daily, so in a real production pipeline, we would almost certainly need to update the pipeline daily to keep up with new records, unless storing daily records, to then load in monthly or quarterly loads. <br> 
Driving Test Centres data is more static, with the UK government releasing a new updated list each quarter. <br>

With all these factors in mind, it seems sensible to still update daily to capture the key records, but understand that much of the reference data will only change quarterly, but most of the FACT data will change & add daily.


### <font color='green'>Approaching Different Scenarios</font>
#### Data Size increased by 100x
The use of AWS redshift allows this pipeline to scale quickly to handle larger data storage requirements within the database for analyst use. This also allows compute to be scaled to handle the users running larger queries. <br>

#### The data is required for a dashboard at 7am each day 
This could be achieved using a tool like Apache Airflow, which can schedule a daily run in the early hours of each morning, to collect the latest data and push that data through the ETL process and into the UK Accidents Database. These tables can then feed the downstream dashboard as required.

#### The database needed to be accessed by 100+ people
AWS redshift again proves useful in this scenario, as we can scale up the cluster(s) for more concurrent use. By running a larger cluster size on the cloud, we can quickly achieve the desired cluster size needed to perform multiple queries by hundreds of analysts accessing data. 
